src/appleseed/renderer/kernel/lighting/imagebasedlighting.cpp

```diff
@@ -113,47 +124,62 @@ void compute_ibl_material_sampling(
         // includes the contribution of a specular component since these are explicitly rejected
         // afterward. We need a mechanism to indicate that we want the contribution of some of
         // the components only.
-        Dual3f incoming;
-        DirectShadingComponents material_value;
-        float material_prob;
-        if (!material_sampler.sample(
-                sampling_context,
-                Dual3d(outgoing),
-                incoming,
-                material_value,
-                material_prob))
+        BSDFSample sample(&shading_point, Dual3f(outgoing));
+        bsdf.sample(
+            sampling_context,
+            bsdf_data,
+            false,              // not adjoint
+            true,               // multiply by |cos(incoming, normal)|
+            bsdf_sampling_modes,
+            sample);
+
+        // Filter scattering modes.
+        if (!(bsdf_sampling_modes & sample.get_mode()))
             continue;
 
         // Discard occluded samples.
         Spectrum transmission;
-        material_sampler.trace_simple(shading_context, incoming.get_value(), transmission);
+        ShadingRay shadow_ray(
+            shading_point.get_point(),
+            Vector3d(sample.m_incoming.get_value()),
+            shading_point.get_ray().m_time,
+            VisibilityFlags::ShadowRay,
+            shading_point.get_ray().m_depth + 1);
+        shadow_ray.m_media = shading_point.get_ray().m_media;
+
+        shading_context.get_tracer().trace_simple(
+            shading_context,
+            shading_point,
+            shadow_ray,
+            transmission);
+
         if (is_zero(transmission))
             continue;
 
-        // Evaluate the environment EDF.
+        // Evaluate the environment's EDF.
```

Comment:
The comment was actually correct :) It's not "the EDF of the environment", it's the "environment EDF" (a specific entity type).